### PR TITLE
hash in supports tuples

### PR DESCRIPTION
**Pull request by max-hoffman**
- improve the hash function to handle tuples recursively
- normalize the left side of the IN comparison so `GetField` is converted to `Literal`
- analyzer will explicitly fail for HASH IN that should convert doesn't (should it still implicitly fall back?)

questions:
- what to do about tuple type conversion? should we store the explicit struct of types and validate as we walk the tree while hashing?
- am i missing any edge cases i need to test?
- add `max_allowed_packet = 67108864` system variable to limit set size?

### CODE REVIEW

File: sql/expression/in.go [hunk 1]
```diff
@@ -240,11 +251,75 @@ func newInMap(expr sql.Expression) (map[uint64]sql.Expression, bool, sql.Type, e
...
+// hashOfTuple will recursively hash a Tuple tree with Literal leaves
+func hashOfTuple(tup Tuple) (uint64, error) {
+	hash := xxhash.New()
+	for _, el := range tup {
+		switch v := el.(type) {
+		case *Literal:
+			if _, err := hash.Write([]byte(fmt.Sprintf("%#v,", v.value))); err != nil {
+				return 0, err
+			}
+		case Tuple:
```
Comment (zachmu) [inline sql/expression/in.go:286]: This is probably more trouble that it's worth. The case we care about is `(col_a, col_b) in ( (1,2), (3,4) )`

Would rip this out and not do this transformation if there is more than a single level of nested tuples. Then see if this can be simplified at all.

**Review by zachmu — APPROVED** [review]
Looks good, would pare back a bit

File: sql/expression/in.go [hunk 2]
```diff
@@ -213,19 +217,26 @@ func (hit *HashInTuple) DebugString() string {
...
 	var t sql.Type
 	switch right := expr.(type) {
 	case Tuple:
 		for _, el := range right {
-			switch l := el.(type) {
-			case *Literal:
-				if t == nil {
-					t = l.Type()
+			if t == nil {
+				t = el.Type()
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/expression/in.go]: originally added to catch heterogeneous upfront. after putting the conversion order back this is unnecessary

File: sql/expression/in.go (hunk 1 above)
Comment (max-hoffman) [reply sql/expression/in.go]: replaces the recursion line with an error, so not a lot shorter, but it might let me do the tuple element casting. will try to add that before merging
